### PR TITLE
Add DeviceNotOnlineException for ServiceClient

### PR DESCRIPTION
**Pull request by jenoola**
## Checklist
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-iot-sdk-csharp/blob/master/.github/CONTRIBUTING.md).
- [X] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- [X] This pull-request is submitted against the `master` branch.

### CODE REVIEW

File: common/src/service/ExceptionHandlingHelper.cs [hunk 1]
```diff
@@ -37,5 +38,21 @@ public static Task<string> GetExceptionMessageAsync(HttpResponseMessage response
...
+            if (Enum.TryParse(httpErrorCode, out errorCode))
+            {
+                switch (errorCode)
+                {
+                    case ErrorCode.DeviceNotOnline:
+                        return new DeviceNotOnlineException(exMsg);
+                }
+            }
+            return new DeviceNotFoundException(exMsg);
+        }
```
Comment (prmathur-microsoft) [inline common/src/service/ExceptionHandlingHelper.cs]: Should it return the generic exception DeviceNotFoundException in all other cases?

Comment (prmathur-microsoft) [reply common/src/service/ExceptionHandlingHelper.cs]: Can you add a test for this scenario as well?

File: common/src/service/ExceptionHandlingHelper.cs [hunk 2]
```diff
@@ -18,7 +19,7 @@ class ExceptionHandlingHelper
             var mappings = new Dictionary<HttpStatusCode, Func<HttpResponseMessage, Task<Exception>>>();
 
             mappings.Add(HttpStatusCode.NoContent, async (response) => new DeviceNotFoundException(await GetExceptionMessageAsync(response).ConfigureAwait(false)));
-            mappings.Add(HttpStatusCode.NotFound, async (response) => new DeviceNotFoundException(await GetExceptionMessageAsync(response).ConfigureAwait(false)));
+            mappings.Add(HttpStatusCode.NotFound, HttpNotFoundMapping);
```
Comment (bikamani) [inline common/src/service/ExceptionHandlingHelper.cs]: Trying to understand why is this done differently now?

File: iothub/service/src/Common/Exceptions/DeviceNotOnlineException.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.Devices.Common.Exceptions
+{
+    [Serializable]
```
Comment (bikamani) [inline iothub/service/src/Common/Exceptions/DeviceNotOnlineException.cs]: Why is this Serializable?

File: iothub/service/src/Common/Exceptions/ErrorCode.cs [hunk 4]
```diff
@@ -40,6 +40,7 @@ public enum ErrorCode
         JobNotFound = 404002,
         PartitionNotFound = 404003,
         ModuleNotFound = 404010,
+        DeviceNotOnline = 404103,
```
Comment (bikamani) [inline iothub/service/src/Common/Exceptions/ErrorCode.cs]: How is this value defined? Is it documented anywhere?

File: common/src/service/ExceptionHandlingHelper.cs (hunk 1 above)
Comment (jenoola) [reply common/src/service/ExceptionHandlingHelper.cs]: It's been returning generic DeviceNotFoundException before, it's just according to exception in http headers. But DeviceNotFound and DeviceNotOnline both have 404 error code, that's why the issue appeared

File: common/src/service/ExceptionHandlingHelper.cs (hunk 2 above)
Comment (jenoola) [reply common/src/service/ExceptionHandlingHelper.cs]: 'DeviceNotFound' and 'DeviceNotOnline' both have 404 error code in http header, so there is only one mapping for both

File: iothub/service/src/Common/Exceptions/ErrorCode.cs (hunk 4 above)
Comment (jenoola) [reply iothub/service/src/Common/Exceptions/ErrorCode.cs]: this code is returned by IoT Hub when device is not online

File: iothub/service/src/Common/Exceptions/DeviceNotOnlineException.cs (hunk 3 above)
Comment (jenoola) [reply iothub/service/src/Common/Exceptions/DeviceNotOnlineException.cs]: this class is implemented as DeviceNotFoundException which is serializable

**Comment by prmathur-microsoft** [comment]
Closing this PR.

**Pull request closed without merging**